consensus/core/src/base_committer.rs

```diff
@@ -0,0 +1,400 @@
+    /// the specified leader (`leader_block`).
+    fn is_vote(&self, potential_vote: &VerifiedBlock, leader_block: &VerifiedBlock) -> bool {
+        let reference = leader_block.reference();
+        let leader_block_slot = BlockSlot::from(reference);
+        self.find_supported_block(leader_block_slot, potential_vote) == Some(reference)
+    }
+
+    /// Check whether the specified block (`potential_certificate`) is a certificate
+    /// for the specified leader (`leader_block`). An `all_votes` map can be
+    /// provided as a cache to quickly skip checking against the block store on
+    /// whether a reference is a vote. This is done for efficiency. Bear in mind
+    /// that the `all_votes` should refer to votes considered to the same `leader_block`
+    /// and it can't be reused for different leaders.
+    fn is_certificate(
+        &self,
+        potential_certificate: &VerifiedBlock,
+        leader_block: &VerifiedBlock,
+        all_votes: &mut HashMap<BlockRef, bool>,
+    ) -> bool {
+        let mut votes_stake_aggregator = StakeAggregator::<QuorumThreshold>::new();
+        for reference in potential_certificate.ancestors() {
+            let is_vote = if let Some(is_vote) = all_votes.get(reference) {
+                *is_vote
+            } else {
+                let potential_vote = self
+                    .dag_state
+                    .read()
+                    .get_block(*reference)
+                    .expect("We should have the whole sub-dag by now");
+                let is_vote = self.is_vote(&potential_vote, leader_block);
+                all_votes.insert(*reference, is_vote);
+                is_vote
+            };
+
+            if is_vote {
+                tracing::trace!("[{self}] {reference} is a vote for {leader_block:?}");
+                if votes_stake_aggregator.add(reference.author, &self.context.committee) {
+                    return true;
+                }
+            } else {
+                tracing::trace!("[{self}] {reference} is not a vote for {leader_block:?}",);
+            }
+        }
+        false
+    }
+
+    /// Decide the status of a target leader from the specified anchor. We commit
+    /// the target leader if it has a certified link to the anchor. Otherwise, we
+    /// skip the target leader.
+    fn decide_leader_from_anchor(
+        &self,
+        anchor: &VerifiedBlock,
+        leader_block_slot: BlockSlot,
+    ) -> LeaderStatus {
+        // Get the block(s) proposed by the leader. There could be more than one leader block
+        // per round (produced by a Byzantine leader).
+        let leader_blocks = self
+            .dag_state
+            .read()
+            .get_blocks_at_block_slot(leader_block_slot);
+
+        // TODO: Remove this check once we have a better way to handle/track Byzantine leaders.
+        let mut seen_rounds = HashSet::new();
+        for block in leader_blocks.iter() {
+            if !seen_rounds.insert(block.round()) {
```

Comment:
Since Slot has a specific round, this loop can be replaced with `!leader_blocks.is_empty()`?

Reply:
Good point think I just have to check `if leader_blocks.len() > 1`